src/controllerclientimpl.cpp

```
@@ -2125,4 +2158,292 @@ void ControllerClientImpl::CreateLogEntries(const std::vector<LogEntry>& logEntr
+                // start the next asynchronous read
+                _ReadFromSubscriptionStream(stream, subscriptionBuffer, onReadHandlers, mutex, allocator);
+                return;
+            }
+        }
+
+        // parse message type
+        if (!result.IsObject() || !result.HasMember("type") || !result["type"].IsString()) {
+            MUJIN_LOG_INFO("receive unexpected websocket message without type field");
+            // start the next asynchronous read
+            _ReadFromSubscriptionStream(stream, subscriptionBuffer, onReadHandlers, mutex, allocator);
+            return;
+        }
+        std::string messageType = result["type"].GetString();
+
+        // ignore pong/ka message
+        if (messageType == "pong" || messageType == "ka") {
+            // start the next asynchronous read
+            _ReadFromSubscriptionStream(stream, subscriptionBuffer, onReadHandlers, mutex, allocator);
+            return;
+        }
+
+        // parse message id
+        if (!result.HasMember("id") || !result["id"].IsString()) {
+            MUJIN_LOG_INFO(boost::format("receive unexpected websocket message without id field of type %s") % messageType);
+            // start the next asynchronous read
+            _ReadFromSubscriptionStream(stream, subscriptionBuffer, onReadHandlers, mutex, allocator);
+            return;
+        }
+        std::string subscriptionId = result["id"].GetString();
+
+        // find the callback function according to subscription id
+        std::unordered_map<std::string, std::function<void(const boost::system::error_code&, rapidjson::Value&&)>>::const_iterator it = onReadHandlers.find(subscriptionId);
+        if (it != onReadHandlers.end() && it->second) {
+            // invoke callback function if there are payloads
+            if (result.HasMember("payload")) {
+                try {
+                    (it->second)(errorCode, std::move(result["payload"]));
```

Can we return the "data" field from the result to the caller instead of the entire "payload"?

---

So I can assume it always returns "data", never "error" or something else?

---

And  can we report the errors in the result via `error_code`?

---

I want to avoid mixing socket-level errors with graphql-level errors.

(It's possible btw)
https://live.boost.org/doc/libs/1_84_0/libs/outcome/doc/html/motivation/plug_error_code2.html

---

Callers of `ExecuteGraphQuery` and `ExecuteGraphQueryRaw` don't deal with extracting data and errors themselves. This is also the case for the python client. I don't see why this API needs an exception. 

> I want to avoid mixing socket-level errors with graphql-level errors.

Why do you want to avoid that?

---

> So I can assume it always returns "data", never "errors" or something else?

You need to check for errors and then for data in the result

---

> Callers of ExecuteGraphQuery and ExecuteGraphQueryRaw don't deal with extracting data and errors

Caller of `ExecuteGraphQueryRaw` does need to parse data/error themselves. The current implementation is consistent with it.

---

So how about just renaming ExecuteGraphSubscription to ExecuteGraphSubscriptionRaw?

---

@LLipter You can also add `ExecuteGraphSubscriptionRaw` but I still want us to support `ExecuteGraphSubscription` with error and data processing, thank you.

---

I see. I updated the callback function accordingly. can you have a look again? @woswos

---

looks good to me, thank you